# Operations Perfomed:
> 1. Reading in the Upper_Secondary school data
> 2. Rescaling the data to prevent data redundancy and to get optimal model performance
> 3. Hypertuning of models' performance for optimal model perfomance
> 4. Model building on the training datasets, evaluation on the testing sets using various regression techniques such as: Linear, Ridge, Lasso, and Gradient boost regression
> 5. Evaluation of models using various metrics in regression such as mean_absolute_error, mean_squared_error root_mean_squared_error, r2_score, residual_sum of squares to evaluate model performance
> 6. Performed comparison on the model between the actual values and the predicted values




In [ ]:
import pandas as pd
import pickle
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
import warnings
warnings.filterwarnings('ignore')


In [ ]:
df= pd.read_csv('/content/Scaled_Upper_Secondary_edu.csv')

In [ ]:
df.head(3)

Year  Completion Rate for upper secondary Edu  Childhood Education GER  \
0  2012                                     96.9                47.348558   
1  2012                                     96.2                47.348558   
2  2013                                     97.0                47.348558   

   Gross enrolment ratio, early childhood educational development programmes  \
0                                           5.871662                           
1                                           5.871662                           
2                                           5.871662                           

   Literacy rate for 25-64 years old  \
0                             68.735   
1                             68.735   
2                             68.735   

   Expenditure on education as a percentage of total government expenditure (%)  \
0                                              19.25                              
1                                              19.25                              
2                                              17.96                              

   Government expenditure on education as a percentage of GDP (%)  \
0                                               3.90                
1                                               3.90                
2                                               3.44                

   Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
0           1.0                        0.0                             0.0   
1           1.0                        0.0                             0.0   
2           1.0                        0.0                             0.0   

   Europe and Northern America  Latin America and the Caribbean  \
0                          0.0                              0.0   
1                          0.0                              0.0   
2                          0.0                              0.0   

   Northern Africa and Western Asia  Oceania  Southern Asia  \
0                               0.0      0.0            0.0   
1                               0.0      0.0            0.0   
2                               0.0      0.0            0.0   

   Sub-Saharan Africa  Gender_numerical  
0                 0.0                 1  
1                 0.0                 0  
2                 0.0                 1

## Data Transformation

#### 1. Data Rescaling

In [ ]:
# Data Rescaling

data_col = df['Year']
df.set_index('Year',inplace =True)

df.head()


Completion Rate for upper secondary Edu  Childhood Education GER  \
Year                                                                     
2012                                     96.9                47.348558   
2012                                     96.2                47.348558   
2013                                     97.0                47.348558   
2013                                     96.4                47.348558   
2014                                     97.2                36.650000   

      Gross enrolment ratio, early childhood educational development programmes  \
Year                                                                              
2012                                           5.871662                           
2012                                           5.871662                           
2013                                           5.871662                           
2013                                           5.871662                           
2014                                           0.000000                           

      Literacy rate for 25-64 years old  \
Year                                      
2012                             68.735   
2012                             68.735   
2013                             68.735   
2013                             68.735   
2014                             68.735   

      Expenditure on education as a percentage of total government expenditure (%)  \
Year                                                                                 
2012                                              19.25                              
2012                                              19.25                              
2013                                              17.96                              
2013                                              17.96                              
2014                                              17.32                              

      Government expenditure on education as a percentage of GDP (%)  \
Year                                                                   
2012                                               3.90                
2012                                               3.90                
2013                                               3.44                
2013                                               3.44                
2014                                               3.42                

      Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
Year                                                                            
2012           1.0                        0.0                             0.0   
2012           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   
2014           1.0                        0.0                             0.0   

      Europe and Northern America  Latin America and the Caribbean  \
Year                                                                 
2012                          0.0                              0.0   
2012                          0.0                              0.0   
2013                          0.0                              0.0   
2013                          0.0                              0.0   
2014                          0.0                              0.0   

      Northern Africa and Western Asia  Oceania  Southern Asia  \
Year                                                             
2012                               0.0      0.0            0.0   
2012                               0.0      0.0            0.0   
2013                               0.0      0.0            0.0   
2013                               0.0      0.0            0.0   
2014                               0.0  

In [ ]:
scaler = MinMaxScaler()
normalized_df = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)
normalized_df['Year'] = data_col
normalized_df.set_index('Year', inplace =True)
normalized_df.head(3)

Completion Rate for upper secondary Edu  Childhood Education GER  \
Year                                                                     
2012                                 0.985625                 0.291174   
2012                                 0.978437                 0.291174   
2013                                 0.986652                 0.291174   

      Gross enrolment ratio, early childhood educational development programmes  \
Year                                                                              
2012                                           0.062772                           
2012                                           0.062772                           
2013                                           0.062772                           

      Literacy rate for 25-64 years old  \
Year                                      
2012                           0.657782   
2012                           0.657782   
2013                           0.657782   

      Expenditure on education as a percentage of total government expenditure (%)  \
Year                                                                                 
2012                                           0.482729                              
2012                                           0.482729                              
2013                                           0.446206                              

      Government expenditure on education as a percentage of GDP (%)  \
Year                                                                   
2012                                           0.286765                
2012                                           0.286765                
2013                                           0.252941                

      Central Asia  Central and Southern Asia  Eastern and South-Eastern Asia  \
Year                                                                            
2012           1.0                        0.0                             0.0   
2012           1.0                        0.0                             0.0   
2013           1.0                        0.0                             0.0   

      Europe and Northern America  Latin America and the Caribbean  \
Year                                                                 
2012                          0.0                              0.0   
2012                          0.0                              0.0   
2013                          0.0                              0.0   

      Northern Africa and Western Asia  Oceania  Southern Asia  \
Year                                                             
2012                               0.0      0.0            0.0   
2012                               0.0      0.0            0.0   
2013                               0.0      0.0            0.0   

      Sub-Saharan Africa  Gender_numerical  
Year                                        
2012                 0.0               1.0  
2012                 0.0               0.0  
2013                 0.0               1.0

In [ ]:
normalized_df.corr().head(3)

Completion Rate for upper secondary Edu  \
Completion Rate for upper secondary Edu                                            1.000000   
Childhood Education GER                                                            0.232376   
Gross enrolment ratio, early childhood educatio...                                 0.350079   

                                                    Childhood Education GER  \
Completion Rate for upper secondary Edu                            0.232376   
Childhood Education GER                                            1.000000   
Gross enrolment ratio, early childhood educatio...                 0.298705   

                                                    Gross enrolment ratio, early childhood educational development programmes  \
Completion Rate for upper secondary Edu                                                      0.350079                           
Childhood Education GER                                                                      0.298705                           
Gross enrolment ratio, early childhood educatio...                                           1.000000                           

                                                    Literacy rate for 25-64 years old  \
Completion Rate for upper secondary Edu                                      0.312730   
Childhood Education GER                                                      0.118609   
Gross enrolment ratio, early childhood educatio...                           0.122515   

                                                    Expenditure on education as a percentage of total government expenditure (%)  \
Completion Rate for upper secondary Edu                                                     -0.055010                              
Childhood Education GER                                                                     -0.182103                              
Gross enrolment ratio, early childhood educatio...                                           0.002370                              

                                                    Government expenditure on education as a percentage of GDP (%)  \
Completion Rate for upper secondary Edu                                                      0.100356                
Childhood Education GER                                                                     -0.017314                
Gross enrolment ratio, early childhood educatio...                                          -0.017604                

                                                    Central Asia  \
Completion Rate for upper secondary Edu                 0.250122   
Childhood Education GER                                -0.064456   
Gross enrolment ratio, early childhood educatio...     -0.008424   

                                                    Central and Southern Asia  \
Completion Rate for upper secondary Edu                              0.046776   
Childhood Education GER                                              0.087568   
Gross enrolment ratio, early childhood educatio...                  -0.093573   

                                                    Eastern and South-Eastern Asia  \
Completion Rate for upper secondary Edu                                   0.064900   
Childhood Education GER                                                   0.157829   
Gross enrolment ratio, early childhood educatio...                        0.091822   

                                                    Europe and Northern America  \
Completion Rate for upper secondary Edu                                0.450880   
Childhood Education GER                                                0.185646   
Gross enrolment ratio, early childhood educatio...                     0.362286   

                                                    Latin America and the Caribbean  \
Completion Rate for upper secondary Edu                                    0.176922   
Childhood Education GER     

In [ ]:
# spliting the dataset into training and testing set
x = normalized_df.drop('Completion Rate for upper secondary Edu', axis = 1)
y = normalized_df['Completion Rate for upper secondary Edu']


# making the train set to be 70% of the entire dataset.
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state=42)


### Parameters hypertuning

In [ ]:
def parameters_hypertuning(model, parameters = {}):
  gs = GridSearchCV(model, parameters)
  gs.fit(x_train, y_train)
  hypertuned_parameters = gs.best_params_

  return hypertuned_parameters


In [ ]:
#hypertuning of paramters for the linear regression model
parameters_hypertuning(linear_model, {'n_jobs': range(0, 50, 5)})

{'n_jobs': 0}

In [ ]:
#hypertuning of paramters for the ridge regression model

parameters_hypertuning(ridge_model,
 {'alpha': np.arange(1.0, 20.0),
  'max_iter': range(0, 1500, 50)
  })

{'alpha': 1.0, 'max_iter': 50}

In [ ]:
#hypertuning of paramters for the lasso regression model
parameters_hypertuning(lasso_model, {'alpha': np.arange(1.0, 20.0),
  'max_iter': range(0, 1500, 50)})

{'alpha': 1.0, 'max_iter': 50}

In [ ]:
#hypertuning of paramters for the Gradient boosting regression model
parameters_hypertuning(Gbr, {'loss': ['squared_error', 'absolute_error', 'huber', 'quantile'],
                             'learning_rate': np.arange(1.0, 20.0, 2.0),
                             'n_estimators': range(0, 50, 5)})

{'learning_rate': 1.0, 'loss': 'squared_error', 'n_estimators': 40}

## Model Training, Testing and evaluation


 Because the target variable been 'Completion rate for primary school Edu' is a continoues data and not categorical. therfore, the model algorithm to be use is the regression and not classifiers.

#### 1. Linear Regression


In [ ]:

import time

# Record start time
start_time = time.time()

# Create a Linear Regression model
linear_model = LinearRegression()

# Fit the model to the training dataset
linear_model.fit(x_train, y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))





Training took 0.04 seconds.


In [ ]:
# Obtain predictions
predicted_values = linear_model.predict(x_test)

#Measuring regression performance
MAE = mean_absolute_error(y_test, predicted_values)
MSE = mean_squared_error(y_test, predicted_values)
RMSE =np.sqrt(mean_squared_error(y_test, predicted_values))
RSS =np.sum(np.square(y_test - predicted_values))
r2=r2_score(y_test, predicted_values)




In [ ]:
# getting the linear regression evaluation results

results = pd.DataFrame([['Linear regression', MAE, MSE, RMSE, RSS,r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results

Model       MAE       MSE      RMSE        RSS  r2_score
0  Linear regression  0.153944  0.039036  0.197575  21.977158   0.55758

In [ ]:
lreg_df = pd.DataFrame(data ={"actual_values": y_test * 100, "predictions":predicted_values * 100})
lreg_df


actual_values  predictions
Year                            
2018      19.601602    47.656250
2012      16.315844    18.554688
2019      84.289968    88.085938
2018      15.289044    51.171875
2014      33.771434    30.078125
...             ...          ...
2013       9.538967    24.609375
2013      50.508266    51.953125
2018      65.766506    54.296875
2015      18.061403    54.492188
2018      73.508574    85.546875

[563 rows x 2 columns]

## Ridge Regression

In [ ]:
# ridge regression
start_time = time.time()

ridge_model = Ridge(alpha = 1.0, max_iter = 50)
ridge_model.fit(x_train, y_train)
# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))


Training took 0.01 seconds.


In [ ]:
y_preds = ridge_model.predict(x_test)

In [ ]:
# getting the ridge model evaluation results
mae= mean_absolute_error(y_test, y_preds)
mse= mean_squared_error(y_test, y_preds)
rmse=np.sqrt(mean_squared_error(y_test, y_preds))
rss=np.sum(np.square(y_test - y_preds))
r_squared=r2_score(y_test, y_preds)

In [ ]:
model_results = pd.DataFrame([['Ridge regression', mae, mse, rmse, rss,r_squared]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)
model_results

Model       MAE       MSE      RMSE        RSS  r2_score
0  Ridge regression  0.153265  0.038952  0.197364  21.930257  0.558524

In [ ]:
ridg_df = pd.DataFrame(data ={"actual_values": y_test * 100, "predictions":predicted_values * 100})
ridg_df


actual_values  predictions
Year                            
2018      19.601602    47.656250
2012      16.315844    18.554688
2019      84.289968    88.085938
2018      15.289044    51.171875
2014      33.771434    30.078125
...             ...          ...
2013       9.538967    24.609375
2013      50.508266    51.953125
2018      65.766506    54.296875
2015      18.061403    54.492188
2018      73.508574    85.546875

[563 rows x 2 columns]

#### Lasso regression

In [ ]:

# Record start time
start_time = time.time()

lasso_model = Lasso(alpha = 1.0, max_iter = 50)

lasso_model.fit(x_train, y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))




Training took 0.02 seconds.


In [ ]:
lasso_predictions  = lasso_model.predict(x_test)

In [ ]:
print(lasso_predictions.shape, y_test.shape)

(563,) (563,)


In [ ]:
# getting the lasso model evaluation results
lasso_r2 = r2_score(y_test, lasso_predictions)
lasso_mse = mean_squared_error( y_test,lasso_predictions)
lasso_rmse = np.sqrt(lasso_mse)
lasso_mae = mean_absolute_error(y_test,lasso_predictions)
lasso_rss = np.sum(np.square(y_test - lasso_predictions))

In [ ]:
model_results = pd.DataFrame([['lasso regression', lasso_mae, lasso_mse, lasso_rmse, lasso_rss, lasso_r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [ ]:
model_results

Model       MAE       MSE      RMSE        RSS  r2_score
0  lasso regression  0.263145  0.088245  0.297061  49.682155 -0.000147

In [ ]:
lasso_df = pd.DataFrame(data ={"actual_values": y_test * 100, "predictions":lasso_predictions * 100})
lasso_df.sort_index(ascending=True, inplace=True)
lasso_df

actual_values  predictions
Year                            
2012      11.181846    47.193455
2012      51.227025    47.193455
2012      98.562481    47.193455
2012      48.742171    47.193455
2012      29.561557    47.193455
...             ...          ...
2021      88.048054    47.193455
2021      88.725742    47.193455
2021      54.307424    47.193455
2021      73.518842    47.193455
2021      12.845261    47.193455

[563 rows x 2 columns]

#### GradientBoostingRegressor

In [ ]:
from sklearn.ensemble import GradientBoostingRegressor


# Record start time
start_time = time.time()

Gbr = GradientBoostingRegressor(learning_rate= 1.0, loss= 'squared_error', n_estimators= 40)
Gbr.fit(x_train,y_train)

# Record end time
end_time = time.time()

# Calculate elapsed time
elapsed_time = end_time - start_time

# Display training time
print("Training took {:.2f} seconds.".format(elapsed_time))



Training took 0.10 seconds.


In [ ]:
Gbr_predictions  = Gbr.predict(x_test)

In [ ]:
# getting the Gradient boosting Model evaluation results
Gbr_r2 = r2_score(y_test, Gbr_predictions)
Gbr_mse = mean_squared_error( y_test,Gbr_predictions)
Gbr_rmse = np.sqrt(Gbr_mse)
Gbr_mae = mean_absolute_error(y_test,Gbr_predictions)
Gbr_rss = np.sum(np.square(y_test - Gbr_predictions))

In [ ]:
model_results = pd.DataFrame([['GBR regression', Gbr_mae, Gbr_mse, Gbr_rmse, Gbr_rss, Gbr_r2]],
               columns = ['Model', 'MAE', 'MSE', 'RMSE', 'RSS','r2_score'])
results = pd.concat([results, model_results], ignore_index=True)

In [ ]:
model_results

Model       MAE       MSE      RMSE       RSS  r2_score
0  GBR regression  0.091208  0.018447  0.135821  10.38587  0.790923

In [ ]:
Gbr_df = pd.DataFrame(data ={"actual_values": y_test * 100, "predictions":lasso_predictions * 100})
Gbr_df.sort_index(ascending=True, inplace=True)
Gbr_df

actual_values  predictions
Year                            
2012      11.181846    47.193455
2012      51.227025    47.193455
2012      98.562481    47.193455
2012      48.742171    47.193455
2012      29.561557    47.193455
...             ...          ...
2021      88.048054    47.193455
2021      88.725742    47.193455
2021      54.307424    47.193455
2021      73.518842    47.193455
2021      12.845261    47.193455

[563 rows x 2 columns]

In [ ]:
# Printing the stored results of all the three models
results.drop_duplicates().head(4)


Model       MAE       MSE      RMSE        RSS  r2_score
0  Linear regression  0.153944  0.039036  0.197575  21.977158  0.557580
1   Ridge regression  0.153265  0.038952  0.197364  21.930257  0.558524
2   lasso regression  0.263145  0.088245  0.297061  49.682155 -0.000147
3     GBR regression  0.091208  0.018447  0.135821  10.385870  0.790923

### conclusion

> Based on the results presented above, it is evident that the Ridge regression model exhibits the most favorable performance. This conclusion is drawn from the fact that the model achieves a mean_absolute_error of 0.1532 which is slightly lower than the linear regression having a mean_absolute error value of 0.1539. This value approaching zero signifies a strong indication of optimal performance within the context of regression modeling.






In [ ]:
def save_model(filename):
  filename=filename
  pickle.dump([linear_model, ridge_model, lasso_model, Gbr], open(filename, 'wb'))

In [ ]:
save_model('Upper_secondary_model.cav')